### PR TITLE
cdc_uart_bridge: Bringup and related changes

### LINKED ISSUE #58: CDC-UART bridge
Implement an uart bridge over usb with no data loss, lag or weird behaviour.

Should also support configuring baudrate, stop bits etc.

### PR DESCRIPTION
**Pull request by kbeckmann**
Closes #58

### CODE REVIEW

**Comment by kbeckmann** [comment]
Did you try it? It seems to work pretty well. The 250ms timeout is there to illustrate the lag, can use 50 or something later.

File: app/cdc_uart_bridge.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
+#include <FreeRTOS.h>
+#include <task.h>
+#include <queue.h>
+
+#include "drivers/max14662.h"
+#include "drivers/uart.h"
+#include "drivers/usb/cdc.h"
+#include "stm32l4xx_hal.h"
+#include "cdc_uart_bridge.h"
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: Move to line 5

File: app/cdc_uart_bridge.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+#define CDC_TX_TASK_NAME			"CDCtx"
+#define CDC_TX_TASK_PRIORITY		1
+
+/* Fun test string: 
+</////////////////////////////////////////////////////////////>[##################]
+<2345678901234567890123456789012345678901234567890123456789012>[ABCDEFGHIJKLMNOPQR]
+*/
+
+#define QUEUE_LENGTH    10
+#define ITEM_SIZE       sizeof(struct rx_queue_item)
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: spaces?

File: app/cdc_uart_bridge.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+static struct {
+	bool initialized;
+	StackType_t cdc_rx_task_stack[CDC_RX_TASK_STACK_SIZE];
+	TaskHandle_t cdc_rx_task_handle;
+	StaticTask_t cdc_rx_task_tcb;
+	StackType_t cdc_tx_task_stack[CDC_RX_TASK_STACK_SIZE];
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: CDC_TX_TASK_STACK_SIZE, or just have 1 define for both

File: app/cdc_uart_bridge.c [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	TaskHandle_t cdc_tx_task_handle;
+	StaticTask_t cdc_tx_task_tcb;
+	StaticQueue_t cdc_tx_queue;
+	QueueHandle_t cdc_tx_queue_handle;
+	uint8_t cdc_tx_queue_storage[QUEUE_LENGTH * ITEM_SIZE];
+	struct rx_queue_item cdc_rx_queue_item;
+} self;
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: Not sure the cdc prefix is needed, also maybe it's possible to move the StackType_t + TaskHandle_t + StaticTask_t into a struct and have 2 structs in the self, then you can do self.rx.task_handle

File: app/cdc_uart_bridge.c [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+static void cdc_rx_task(void *p_arg)
+{
+	err_t r;
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: You can probably move cdc_rx_queue_item from self to here.

File: app/cdc_uart_bridge.c [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+static void cdc_tx_task(void *p_arg)
+{
+	err_t r;
+	struct rx_queue_item item;
+
+	for (;;) {
+		if (xQueueReceive(self.cdc_tx_queue_handle, &item, 250) == pdFALSE) {
+			uart_flush_rx();
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: Maybe a comment why we do this.

File: app/cdc_uart_bridge.c (hunk 2 above)
Comment (arturo182) [inline app/cdc_uart_bridge.c]: QUEUE_ITEM_SIZE

File: app/cdc_uart_bridge.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		CDC_TX_TASK_PRIORITY,
+		&self.cdc_tx_task_stack[0],
+		&self.cdc_tx_task_tcb);
+	if (self.cdc_tx_task_handle == NULL)
+		return ECDC_UART_BRIDGE_TASK_CREATE;
+
+	self.cdc_tx_queue_handle = xQueueCreateStatic(QUEUE_LENGTH,
+								ITEM_SIZE,
+								self.cdc_tx_queue_storage,
+								&self.cdc_tx_queue);
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: Inconsistent indent with the above xTaskCreateStatic, I think one extra indent to the original call is good.

File: app/cdc_uart_bridge.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+	// TODO: enable only rx and tx
+	max14662_set_value(MAX14662_AD_0_0, 0xff);
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: check return value?

File: app/cdc_uart_bridge.c [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+	self.initialized = true;
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: Move to 117

File: app/cdc_uart_bridge.c [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+	r = uart_start_rx(self.cdc_tx_queue_handle);
```
Comment (arturo182) [inline app/cdc_uart_bridge.c]: Check r so we don't set initialized = true if starting fails

File: app/drivers/pcd.c [hunk 11]
```diff
@@ -131,7 +131,7 @@ void HAL_PCD_DataInStageCallback(PCD_HandleTypeDef *p_pcd, uint8_t epnum)
 	} else {
 		for (int i = 0; i < USBD_MAX_NUM_CLASSES; ++i) {
 			if ((self.p_usbd->pClasses[i]->DataIn != NULL) && (self.p_usbd->dev_state == USBD_STATE_CONFIGURED))
-				self.p_usbd->pClasses[i]->DataIn(self.p_usbd, epnum);
+				self.p_usbd->pClasses[i]->DataIn(self.p_usbd, epnum | 0x80);
```
Comment (arturo182) [inline app/drivers/pcd.c]: I thought this was already merged.

File: app/drivers/uart.c [hunk 12]
```diff
@@ -3,15 +3,21 @@
 #include <queue.h>
 #include <task.h>
 
+#include "stm32l4xx_hal.h"
+#include "stm32l4xx_ll_dma.h"
```
Comment (arturo182) [inline app/drivers/uart.c]: Move to line 12

File: app/drivers/uart.c [hunk 13]
```diff
@@ -3,15 +3,21 @@
...
 #include "drivers/uart.h"
+#include "drivers/usb/cdc.h"
 
 static struct {
 	DMA_HandleTypeDef hdma_usart3_tx;
 	DMA_HandleTypeDef hdma_usart3_rx;
 	UART_HandleTypeDef uart_handle;
 
+	struct rx_queue_item uart_rx_item;
+	QueueHandle_t uart_rx_queue_handle;
```
Comment (arturo182) [inline app/drivers/uart.c]: Don't think the uart prefix is needed (2x)

File: app/drivers/uart.c [hunk 14]
```diff
@@ -116,8 +122,23 @@ void HAL_UART_RxCpltCallback(UART_HandleTypeDef *huart)
 {
 	static BaseType_t xHigherPriorityTaskWoken = pdFALSE;
 
-	xSemaphoreGiveFromISR(self.rx_done_semaphore, &xHigherPriorityTaskWoken);
+	xQueueSendFromISR(self.uart_rx_queue_handle, &self.uart_rx_item, &xHigherPriorityTaskWoken);
 	portYIELD_FROM_ISR(xHigherPriorityTaskWoken);
+	HAL_UART_Receive_DMA(&self.uart_handle, (uint8_t*) self.uart_rx_item.data, USB_FS_MAX_PACKET_SIZE);
+}
+
+void HAL_UART_AbortReceiveCpltCallback (UART_HandleTypeDef *huart)
```
Comment (arturo182) [inline app/drivers/uart.c]: Remove the space before "(".

Also we should rename huart to p_uart in this file.

File: app/drivers/uart.c [hunk 15]
```diff
@@ -170,36 +191,26 @@ err_t uart_tx(const uint8_t *p_buf, uint32_t size, uint32_t timeout_ticks, bool
...
-	if (xSemaphoreTake(self.rx_done_semaphore, timeout_ticks) == pdFALSE) {
-		status = HAL_UART_AbortReceive(&self.uart_handle);
-		/* Give semaphore because the IRQ handler will never be called if we abort */
-		xSemaphoreGive(self.rx_done_semaphore);
-		r = EUART_RX_TIMEOUT;
-		goto out;
-	}
-	xSemaphoreGive(self.rx_done_semaphore);
+	/* HAL_UART_AbortReceiveCpltCallback restarts DMA */
+	status = HAL_UART_AbortReceive_IT(&self.uart_handle);
```
Comment (arturo182) [inline app/drivers/uart.c]: check status?

File: app/drivers/uart.h [hunk 16]
```diff
@@ -1,6 +1,8 @@
 #pragma once
 
 #include <stdbool.h>
+#include "FreeRTOS.h"
+#include "queue.h"
```
Comment (arturo182) [inline app/drivers/uart.h]: Chane to < > (2x) and new line between < and " includes

File: app/drivers/usb/cdc.c [hunk 17]
```diff
@@ -5,23 +5,28 @@
 
 #include <FreeRTOS.h>
 #include <semphr.h>
+#include <queue.h>
 #include <stdbool.h>
 
-#define CLASS_IDX						1
+#define CLASS_IDX		1
+#define QUEUE_LENGTH	10
+#define ITEM_SIZE		sizeof(struct rx_queue_item)
```
Comment (arturo182) [inline app/drivers/usb/cdc.c]: QUEUE_ITEM_SIZE

File: app/drivers/usb/cdc.c [hunk 18]
```diff
@@ -5,23 +5,28 @@
...
 
 static struct {
 	bool initialized;
 	USBD_HandleTypeDef *p_usbd;
 	PCD_HandleTypeDef *p_pcd;
 	uint8_t rx_buff[USB_FS_MAX_PACKET_SIZE];
 	uint8_t ctrl_buff[USB_FS_MAX_PACKET_SIZE];
-	uint8_t rx_len;
-	volatile bool tx_busy;
+	struct rx_queue_item rx_buf;
```
Comment (arturo182) [inline app/drivers/usb/cdc.c]: rx_buff and rx_buf in same struct is not good.

File: app/drivers/usb/cdc.c [hunk 19]
```diff
@@ -172,11 +178,10 @@ static uint8_t cdc_data_out(USBD_HandleTypeDef *p_dev, uint8_t epnum)
...
 
-	self.rx_len = HAL_PCD_EP_GetRxCount(self.p_pcd, epnum);
-
-	status = HAL_PCD_EP_Receive(self.p_pcd, CDC_OUT_EP, self.rx_buff, USB_FS_MAX_PACKET_SIZE);
-
-	xSemaphoreGiveFromISR(self.rx_done_semaphore, &xHigherPriorityTaskWoken);
+	self.rx_buf.len = HAL_PCD_EP_GetRxCount(self.p_pcd, epnum);
+	status = HAL_PCD_EP_Receive(self.p_pcd, CDC_OUT_EP, self.rx_buf.data, USB_FS_MAX_PACKET_SIZE);
+ 
+	xQueueSendFromISR(self.rx_queue_handle, &self.rx_buf, &xHigherPriorityTaskWoken);
```
Comment (arturo182) [inline app/drivers/usb/cdc.c]: So like I said, the HAL_PCD_EP_Receive sets up the buffer for _next_ transfer, I think we should do the queue send before the HAL_PCD_EP_Receive call.

File: app/drivers/usb/cdc.c [hunk 20]
```diff
@@ -251,7 +250,13 @@ err_t usb_cdc_init(USBD_HandleTypeDef *p_usbd, PCD_HandleTypeDef *p_pcd)
 	if (status != HAL_OK)
 		return EUSB_CDC_REG_CLASS;
 
-	self.rx_done_semaphore = xSemaphoreCreateBinaryStatic(&self.rx_done_semaphore_buffer);
+	self.rx_queue_handle = xQueueCreateStatic(QUEUE_LENGTH,
+							ITEM_SIZE,
+							self.rx_queue_storage,
+							&self.rx_queue);
```
Comment (arturo182) [inline app/drivers/usb/cdc.c]: one extra indent in addition to line 253 is enough

File: app/drivers/usb/cdc.h [hunk 21]
```diff
@@ -15,6 +15,11 @@
 #define CDC_IN_EP				0x82
 #define CDC_OUT_EP				0x02
 
-err_t usb_cdc_rx(const uint8_t **pp_buf, uint16_t *p_len, uint32_t timeout_ticks);
+struct rx_queue_item {
+	uint8_t len;
+	uint8_t data[64];
```
Comment (arturo182) [inline app/drivers/usb/cdc.h]: use USB_FS_MAX_PACKET_SIZE

File: app/errors.h [hunk 22]
```diff
@@ -19,16 +19,17 @@ typedef uint32_t err_t;
...
+#define EI2C_BASE				(ERR_BASE + (0x03 << 16))
+#define ELED_BASE				(ERR_BASE + (0x04 << 16))
+#define EMAX14662_BASE			(ERR_BASE + (0x05 << 16))
+#define EMCP4018T_BASE			(ERR_BASE + (0x06 << 16))
+#define EUART_BASE				(ERR_BASE + (0x07 << 16))
+#define EPOWER_BASE				(ERR_BASE + (0x08 << 16))
+#define EUSB_BASE				(ERR_BASE + (0x09 << 16))
+#define EUSB_CDC_BASE			(ERR_BASE + (0x0A << 16))
+#define EUSB_HID_BASE			(ERR_BASE + (0x0B << 16))
+#define ECDC_UART_BRIDGE_BASE	(ERR_BASE + (0x0C << 16))
```
Comment (arturo182) [inline app/errors.h]: Man I hate github formatting

Comment (arturo182) [reply app/errors.h]: No more! https://github.com/sindresorhus/tab-size-on-github

**Comment by kbeckmann** [comment]
:sweat_smile:

File: app/cdc_uart_bridge.c (hunk 9 above)
Comment (kbeckmann) [reply app/cdc_uart_bridge.c]: i don't understand this one

File: app/drivers/usb/cdc.c (hunk 19 above)
Comment (kbeckmann) [reply app/drivers/usb/cdc.c]: If I move it, it doesn't work well anymore. Not sure exactly how this works tbh.

File: app/cdc_uart_bridge.c [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+static void tx_task(void *p_arg)
+{
+	err_t r;
+	struct rx_queue_item item;
+
+	for (;;) {
+		/* uart_start_rx starts a continuous DMA transfer of 64 bytes that sends its data to our 
+		 * queue that we receive here. In order to get shorter messages in near realtime we need
+		 * a timeout and flush the received bytes so far. */
+		if (xQueueReceive(self.tx_queue_handle, &item, 250) == pdFALSE) {
```
Comment (kbeckmann) [inline app/cdc_uart_bridge.c]: self.tx_queue_handle hmm. confusing name now without a prefix. it's the tx task's queue but it is receiving uart data

**Pull request merged**